File: src/state/allocator.ts
import type { Config } from "../config/config";
import { allocationPorts } from "./state";
import type { Allocation, State } from "./state";

export function findNextAvailablePorts(
  state: State,
  config: Config,
  blockSize: number,
): number[] | null {
  const used = new Set<number>(config.excludedPorts);
  for (const a of state.allocations) {
    for (const p of allocationPorts(a)) used.add(p);
  }

  const candidates: number[] = [];
  for (
    let basePort = config.portRangeStart;
    basePort + blockSize - 1 <= config.portRangeEnd;
    basePort += 1
  ) {
    let allAvailable = true;
    for (let i = 0; i < blockSize; i++) {
      if (used.has(basePort + i)) {
        allAvailable = false;
        break;
      }
    }
    if (allAvailable) candidates.push(basePort);
  }

  if (candidates.length === 0) return null;

  const basePort = candidates[Math.floor(Math.random() * candidates.length)];
  const result: number[] = [];
  for (let i = 0; i < blockSize; i++) result.push(basePort + i);
  return result;
}

export function findLeastRecentlyUsed(
  state: State,
  predicate?: (a: Allocation) => boolean,
): Allocation | null {
  const candidates = predicate
    ? state.allocations.filter(predicate)
    : state.allocations;
  if (candidates.length === 0) return null;
  return candidates.reduce((oldest, current) =>
    current.timestamp < oldest.timestamp ? current : oldest,
  );
}

File: internal/state/allocator.go
package state

import (
	"math/rand"

	"github.com/sylophi/port-pool/internal/config"
)

// FindNextAvailablePorts scans the configured pool range for a contiguous
// block of `blockSize` ports not in use (excluded or allocated). Returns
// nil if no such block exists.
//
// Picks a *random* valid base port rather than the first-fit, on purpose:
// processes outside port-pool's tracking (one-off scripts, system services
// bound to unknown ports inside the range) can still occupy ports we
// don't know about. Random spreading reduces the probability that an
// allocation collides with such an external listener compared to always
// returning the lowest available block.
//
// The math/rand source is package-global and auto-seeded by the Go
// runtime; no explicit seed is needed.
func FindNextAvailablePorts(s *State, cfg *config.PoolConfig, blockSize int) []int {
	used := make(map[int]struct{}, len(cfg.ExcludedPorts)+len(s.Allocations)*blockSize)
	for _, p := range cfg.ExcludedPorts {
		used[p] = struct{}{}
	}
	for _, a := range s.Allocations {
		for _, p := range a.Ports {
			used[p] = struct{}{}
		}
	}

	var candidates []int
	for base := cfg.PortRangeStart; base+blockSize-1 <= cfg.PortRangeEnd; base++ {
		ok := true
		for i := 0; i < blockSize; i++ {
			if _, taken := used[base+i]; taken {
				ok = false
				break
			}
		}
		if ok {
			candidates = append(candidates, base)
		}
	}

	if len(candidates) == 0 {
		return nil
	}
	base := candidates[rand.Intn(len(candidates))]
	out := make([]int, blockSize)
	for i := 0; i < blockSize; i++ {
		out[i] = base + i
	}
	return out
}

// FindLeastRecentlyUsed returns the allocation matching pred with the oldest
// timestamp, or nil if none match. Pass nil pred to consider all allocations.
func FindLeastRecentlyUsed(s *State, pred func(*Allocation) bool) *Allocation {
	var oldest *Allocation
	for i := range s.Allocations {
		a := &s.Allocations[i]
		if pred != nil && !pred(a) {
			continue
		}
		if oldest == nil || a.Timestamp < oldest.Timestamp {
			oldest = a
		}
	}
	return oldest
}
